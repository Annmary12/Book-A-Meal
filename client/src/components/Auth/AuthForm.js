import React, { Component } from 'react';
import PropTypes from 'prop-types';
import InputField from '../Form/InputField.js';
import Loader from '../Loader.js';
import styles from './auth.scss';

/**
 * @class AuthForm
 */
class AuthForm extends Component {
  submitBtnTxt = () => {
    switch (this.props.type) {
      case 'login':
        return 'Login';
      case 'signup':
        return 'Sign Up';
      case 'reset':
        return 'Send reset password email';
      case 'changepassword':
        return 'Change password';
      default:
        return 'Sign Up';
    }
  }

  isSignUp = () => this.props.type === 'signup';

  isLogin = () => this.props.type === 'login';

  isReset = () => this.props.type === 'reset';

  isChangePassword = () => this.props.type === 'changepassword';

  /**
   * @return {undefined} - undefined
   */
  render() {
    const {
      values,
      touched,
      errors,
      isSubmitting,
      handleChange,
      handleSubmit
    } = this.props;
    return (
      <form onSubmit={handleSubmit}>
        {
          this.isSignUp() &&
          <InputField name="fullName" label="Full name" {...this.props} />
        }
        {
          (this.isSignUp() || this.isLogin()) &&
          <InputField name="username" label="Username" {...this.props} />
        }
        {
          (this.isSignUp() || this.isReset()) &&
          <InputField name="email" label="Email" {...this.props} type="email" />
        }

        {
          this.isSignUp()
          &&
          <div>
            Role:
            <div>
              <label>
                Customer
                <input
                  type="radio"
                  value="customer"
                  name="role"
                  className={styles.roleCheckbox}
                  onChange={handleChange}
                  checked={values.role === 'customer' ? 'checked' : ''}
                />
              </label>
              <label>
                Caterer
                <input
                  type="radio"
                  value="caterer"
                  name="role"
                  className={styles.roleCheckbox}
                  onChange={handleChange}
                  checked={values.role === 'caterer' ? 'checked' : ''}
                />
              </label>
            </div>
            {
              errors.role &&
              touched.role &&
              <div className={styles.errorMessage}>{errors.role}</div>
            }
          </div>
        }

        {
          !this.isReset() &&
          <InputField name="password" label="Password" {...this.props} type="password" />
        }

        {

          (this.isSignUp() || this.isChangePassword()) &&
          <InputField name="passwordConfirm" label="Confirm password" {...this.props} type="password" />
        }
        <button
          className={isSubmitting ? styles.loginBtnSubmit : styles.loginBtn}
          type="submit"
          onClick={handleSubmit}
          disabled={isSubmitting}
        >
          {
            isSubmitting ?
              <Loader message="" width={30} /> :
              this.submitBtnTxt()
          }
        </button>
      </form>
    );
  }
}

AuthForm.propTypes = {
  type: PropTypes.string,
  values: PropTypes.objectOf(PropTypes.string),
  touched: PropTypes.objectOf(PropTypes.bool),
  errors: PropTypes.objectOf(PropTypes.string),
  isSubmitting: PropTypes.bool,
  handleChange: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired
};

AuthForm.defaultProps = {
  type: 'signup',
  values: {},
  touched: {},
  errors: {},
  isSubmitting: false
};

export default AuthForm;
