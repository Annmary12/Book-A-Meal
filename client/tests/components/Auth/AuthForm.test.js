import React from 'react';
import AuthForm from '../../../src/components/Auth/AuthForm';


describe('AuthForm component', () => {
  const props = {
    handleChange: jest.fn(),
    handleSubmit: jest.fn()
  }

  it('should match snapshot if type prop is login', () => {
    const wrapper = shallow(
      <AuthForm
        type="login"
        {...props}
      />
    );

    expect(wrapper).toMatchSnapshot();
  })
  it('should match snapshot if type prop is signup', () => {
    const wrapper = shallow(
      <AuthForm
        type="signup"
        {...props}
      />
    );

    expect(wrapper).toMatchSnapshot();
  })
  it('should match snapshot if type prop is reset', () => {
    const wrapper = shallow(
      <AuthForm
        type="reset"
        {...props}
      />
    );

    expect(wrapper).toMatchSnapshot();
  })
  it('should match snapshot if type prop is changepassword', () => {
    const wrapper = shallow(
      <AuthForm
        type="changepassword"
        {...props}
      />
    );

    expect(wrapper).toMatchSnapshot();
  })

  it('should match snapshot if type prop not given or is not valid', () => {
    const wrapper = shallow(
      <AuthForm
        type="test"
        {...props}
      />
    );

    expect(wrapper).toMatchSnapshot();
  })
})