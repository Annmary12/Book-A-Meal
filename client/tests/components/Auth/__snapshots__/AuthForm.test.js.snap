// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm component should match snapshot if type prop is changepassword 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <InputField
    errors={Object {}}
    handleChange={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Password"
    name="password"
    touched={Object {}}
    type="password"
    values={Object {}}
  />
  <InputField
    errors={Object {}}
    handleChange={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Confirm password"
    name="passwordConfirm"
    touched={Object {}}
    type="password"
    values={Object {}}
  />
  <button
    className="loginBtn"
    disabled={false}
    onClick={[MockFunction]}
    type="submit"
  >
    Change password
  </button>
</form>
`;

exports[`AuthForm component should match snapshot if type prop is login 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <InputField
    errors={Object {}}
    handleChange={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Username"
    name="username"
    touched={Object {}}
    type="login"
    values={Object {}}
  />
  <InputField
    errors={Object {}}
    handleChange={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Password"
    name="password"
    touched={Object {}}
    type="password"
    values={Object {}}
  />
  <button
    className="loginBtn"
    disabled={false}
    onClick={[MockFunction]}
    type="submit"
  >
    Login
  </button>
</form>
`;

exports[`AuthForm component should match snapshot if type prop is reset 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <InputField
    errors={Object {}}
    handleChange={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Email"
    name="email"
    touched={Object {}}
    type="email"
    values={Object {}}
  />
  <button
    className="loginBtn"
    disabled={false}
    onClick={[MockFunction]}
    type="submit"
  >
    Send reset password email
  </button>
</form>
`;

exports[`AuthForm component should match snapshot if type prop is signup 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <InputField
    errors={Object {}}
    handleChange={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Full name"
    name="fullName"
    touched={Object {}}
    type="signup"
    values={Object {}}
  />
  <InputField
    errors={Object {}}
    handleChange={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Username"
    name="username"
    touched={Object {}}
    type="signup"
    values={Object {}}
  />
  <InputField
    errors={Object {}}
    handleChange={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Email"
    name="email"
    touched={Object {}}
    type="email"
    values={Object {}}
  />
  <div>
    Role:
    <div>
      <label>
        Customer
        <input
          checked=""
          className="roleCheckbox"
          name="role"
          onChange={[MockFunction]}
          type="radio"
          value="customer"
        />
      </label>
      <label>
        Caterer
        <input
          checked=""
          className="roleCheckbox"
          name="role"
          onChange={[MockFunction]}
          type="radio"
          value="caterer"
        />
      </label>
    </div>
  </div>
  <InputField
    errors={Object {}}
    handleChange={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Password"
    name="password"
    touched={Object {}}
    type="password"
    values={Object {}}
  />
  <InputField
    errors={Object {}}
    handleChange={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Confirm password"
    name="passwordConfirm"
    touched={Object {}}
    type="password"
    values={Object {}}
  />
  <button
    className="loginBtn"
    disabled={false}
    onClick={[MockFunction]}
    type="submit"
  >
    Sign Up
  </button>
</form>
`;

exports[`AuthForm component should match snapshot if type prop not given or is not valid 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <InputField
    errors={Object {}}
    handleChange={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Password"
    name="password"
    touched={Object {}}
    type="password"
    values={Object {}}
  />
  <button
    className="loginBtn"
    disabled={false}
    onClick={[MockFunction]}
    type="submit"
  >
    Sign Up
  </button>
</form>
`;
